src/trystero-handlers.js

```diff
@@ -0,0 +1,250 @@
+  farmhand.updatePeer(peerId, peerState)
+}
+
+/**
+ * Handles another player's initiation of a cow trade request.
+ * @param {Farmhand} farmhand
+ * @param {Object} cowTradeRequestPayload
+ * @param {farmhand.cow} cowTradeRequestPayload.cowOffered
+ * @param {farmhand.cow} cowTradeRequestPayload.cowRequested
+ * @param {string} peerId
+ */
+export const handleCowTradeRequest = async (
+  farmhand,
+  { cowOffered, cowRequested },
+  peerId
+) => {
+  let wasTradeSuccessful = false
+
+  farmhand.setState(
+    state => {
+      const {
+        allowCustomPeerCowNames,
+        cowIdOfferedForTrade,
+        cowsTraded,
+        cowInventory,
+        id,
+        isAwaitingCowTradeRequest,
+        peers,
+        sendCowAccept,
+        sendCowReject,
+      } = state
+
+      if (!sendCowAccept || !sendCowReject) {
+        console.error('Peer connection not set up correctly')
+        return null
+      }
+
+      const cowToTradeAway = cowInventory.find(
+        ({ id }) => id === cowIdOfferedForTrade
+      )
+
+      if (
+        isAwaitingCowTradeRequest ||
+        cowRequested.id !== cowIdOfferedForTrade ||
+        !cowToTradeAway
+      ) {
+        sendCowReject(
+          {
+            reason: cowTradeRejectionReason.REQUESTED_COW_UNAVAILABLE,
+          },
+          peerId
+        )
+
+        return null
+      }
+
+      const updatedCowOffered = {
+        ...cowOffered,
+        timesTraded:
+          cowOffered.originalOwnerId === id
+            ? cowOffered.timesTraded
+            : cowOffered.timesTraded + 1,
+      }
+
+      const [, peerMetadata] = Object.entries(peers).find(
+        ([, { id }]) => id === updatedCowOffered.ownerId
+      )
+
+      state = changeCowAutomaticHugState(state, cowToTradeAway, false)
+      state = removeCowFromInventory(state, cowToTradeAway)
+      state = addCowToInventory(state, {
+        ...updatedCowOffered,
+        ownerId: id,
+      })
+      state = showNotification(
+        state,
+        COW_TRADED_NOTIFICATION`${cowToTradeAway}${cowOffered}${id}${allowCustomPeerCowNames}`,
+        'success'
+      )
+
+      sendCowAccept({ ...cowToTradeAway, isUsingHuggingMachine: false }, peerId)
+
+      wasTradeSuccessful = true
+
+      return {
+        ...state,
+        cowIdOfferedForTrade: updatedCowOffered.id,
+        cowsTraded:
+          updatedCowOffered.originalOwnerId === id
+            ? cowsTraded
+            : cowsTraded + 1,
+        isAwaitingCowTradeRequest: true,
+        selectedCowId: updatedCowOffered.id,
+        peers: {
+          ...peers,
+          [peerId]: {
+            ...peerMetadata,
+            cowOfferedForTrade: { ...cowToTradeAway, ownerId: peerMetadata.id },
+          },
+        },
+      }
+    },
+    async () => {
+      if (!wasTradeSuccessful) return
+
+      // Allow any pending state updates to complete so that an invalid state
+      // is not persisted
+      await sleep(750)
```

Comment:
what's the reason it can be lagging behind so much? i feel obliged to ask if there's anything else that can be done here, though it's mostly just to raise the question. like, is there some sort of event or callback approach instead?